### PR TITLE
NXP UCANS32K PWM support continued

### PR DESCRIPTION
**Pull request by davids5**
Continuation of https://github.com/PX4/Firmware/pull/14978

@jarivanewijk - Thank you for doing all the heavy lifting on this! I have rebased and continued the PR here.

image

image

### CODE REVIEW

**Comment by davids5** [comment]
fyi @igalloway , @jarivanewijk

**Comment by igalloway** [comment]
FYI - I've pinged internally to request review

**Comment by jarivanewijk** [comment]
Thanks David!

I'm trying to reproduce your test with the same command, but I'm unable to get any PWM output. Which FlexTimer and channel were you using? And what pin did you use as output? I tried with FTM0, channel 7, pin E9 (labeled SPI_INT_N pin in the schematics). That's also what is already in timer_config.cpp. I had to set the pin multiplexing manually with a debugger.

Note that at some point we will have to put in the right FlexTimer, channel and pin, as this was just done for testing with an older revision of the board. We're still waiting to receive the production version of the board here in the Netherlands. On the production version I think it should be FTM2, CH1, pin PTA0.

**Comment by davids5** [comment]
> Thanks David!
> 
> I'm trying to reproduce your test with the same command, but I'm unable to get any PWM output. Which FlexTimer and channel were you using? And what pin did you use as output? I tried with FTM0, channel 7, pin E9 (labeled SPI_INT_N pin in the schematics). That's also what is already in timer_config.cpp. I had to set the pin multiplexing manually with a debugger. Is there anything else that I need to configure first?
> 
> Note that at some point we will have to put in the right FlexTimer, channel and pin, as this was just done for testing with an older revision of the board. We're still waiting to receive the production version of the board here in the Netherlands. On the production version I think it should be FTM2, CH1, pin PTA0.

Sorry , you need?
```
pwm arm
```

**Comment by davids5** [comment]
>Note that at some point we will have to put in the right FlexTimer, channel and pin, as this was just done for testing with an older revision of the board. We're still waiting to receive the production version of the board here in the Netherlands. On the production version I think it should be FTM2, CH1, pin PTA0.

That would will just be a two line change in timer_config,

No other changes should be needed.

**Comment by jarivanewijk** [comment]
> Sorry , you need.
> 
> ```
> pwm arm
> ```
> 
> Connected to P1-6

... Of course... Works now. Thank you. ;)

File: platforms/nuttx/src/px4/nxp/s32k1xx/io_pins/io_timer.c [hunk 1]
```diff
@@ -641,6 +639,14 @@ int io_timer_init_timer(unsigned timer)
 		if (handler) {
 			irq_attach(io_timers[timer].vectorno, handler, NULL);
 			up_enable_irq(io_timers[timer].vectorno);
+#if defined(CONFIG_ARCH_CHIP_S32K14X) /* Quick (and not very nice) workaround for the multiple interrupts per FTM on S32K14X */
+			irq_attach(io_timers[timer].vectorno + 1, handler, NULL);
```
Comment (jarivanewijk) [inline platforms/nuttx/src/px4/nxp/s32k1xx/io_pins/io_timer.c:643]: I'm still not really happy about this quick workaround. On S32K14X each FlexTimer has multiple interrupts (While S32K11X only has one per FTM). I just register and enable all of them, but that's probably not something you want to do if you're using the same FlexTimer for other purposes. We can just leave it until we need to change it, but I'm open to suggestions to improve this.

File: platforms/nuttx/src/px4/nxp/s32k1xx/include/px4_arch/io_timer_hw_description.h [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+static inline constexpr timer_io_channels_t initIOTimerChannel(const io_timers_t io_timers_conf[MAX_IO_TIMERS],
+		Timer::TimerChannel timer, GPIO::GPIOPin pin)
+{
+	timer_io_channels_t ret{};
+
+	uint32_t gpio_af = 0;
+
+	switch (pin.port) {
+	case GPIO::PortA:
+		if (pin.pin == GPIO::Pin0 || pin.pin == GPIO::Pin1 || pin.pin == GPIO::Pin2 || pin.pin == GPIO::Pin3 ||
```
Comment (jarivanewijk) [inline platforms/nuttx/src/px4/nxp/s32k1xx/include/px4_arch/io_timer_hw_description.h:57]: This monstrosity should work, but I'm open to suggestions to improve this...

File: platforms/nuttx/src/px4/nxp/s32k1xx/io_pins/io_timer.c (hunk 1 above)
Comment (davids5) [reply platforms/nuttx/src/px4/nxp/s32k1xx/io_pins/io_timer.c:643]: The interrupts are only used for Capture. Which is not implemented. https://github.com/PX4/Firmware/blob/a4f7b1741024f57649f2e9f86381c949b4b661ec/platforms/nuttx/src/px4/nxp/s32k1xx/io_pins/io_timer.c#L151

In the io_timer_set_enable we do this, by timer or channel. But I would have to be clear on what the HW difference is.

File: platforms/nuttx/src/px4/nxp/s32k1xx/include/px4_arch/io_timer_hw_description.h (hunk 2 above)
Comment (davids5) [reply platforms/nuttx/src/px4/nxp/s32k1xx/include/px4_arch/io_timer_hw_description.h:57]: It is all compile time, ugly but necessary, you could format it format it one per line:
```
pin.pin == GPIO::Pin0 || 
pin.pin == GPIO::Pin1 || 
pin.pin == GPIO::Pin2...
```

**Comment by davids5** [comment]
@jarivanewijk - There is a CI failure on the K66, not related to this PR. Looks like we are out of RAM (or it is corrupted) it is on my to-do list. We are probably going to take another update from Upstream Nuttx to resolve some of the issues,

**Comment by davids5** [comment]
@igalloway @jarivanewijk I have added the PWM LED support and rebased on current master.

**Comment by igalloway** [comment]
> @igalloway @jarivanewijk I have added the PWM LED support and rebased on current master.

Thanks. Landon Haugh is checking this. We expect it to be fine.

**Comment by igalloway** [comment]
The RGB PWM was tested and is working.

**Review by TSC21 — APPROVED** [review]
LGTM

**Pull request merged**